### PR TITLE
Fine-tune, validate and evaluate a model

### PR DESCRIPTION
**Pull request by kimground**

### CODE REVIEW

File: conf.d/config.yaml.example [hunk 1]
```diff
@@ -0,0 +1,27 @@
+data:
```
Comment (yj0604park) [inline conf.d/config.yaml.example:1]: 그냥 궁금한거긴 한데 yaml이 널리 쓰이는 중인건가요? 확실히 읽고 고치기에는 json보다는 편하긴 할듯?

File: conf.d/config.yaml.example [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  name: vit_large_patch16_224_in21k
+
+train:
+  batch: 512
+  epochs: 5
+  learning_rate: 0.01
+  log_steps: 100    # loss will be logged every log_steps. This value should not be 0
+
+valid:
+  whether: False
```
Comment (yj0604park) [inline conf.d/config.yaml.example:18]: 여기서 whether는 무슨 의미로 쓰인걸까요?! 코드 보니깐 학습중에 eval 하나 안하나 인거 같긴 한데

File: main.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    print('Load data ...')  # TODO: Modify hard code for dataset
+    train_data = datasets.CIFAR100(root=data_dir, train=True, download=True, transform=transforms.Compose(
+        [transforms.ToTensor(), transforms.Normalize(norm['mean'], norm['stdev'])]
+    ))
+
+    print('initialize model ...')
+    model = ClsModel(config['model']['name'], config['device'], img_size=config['data']['img_size'])
+
+    # TODO: Modify hard code for optimizer and loss function
+    optimizer = optim.SGD(model.parameters(), lr=config['train']['learning_rate'], momentum=0.9, weight_decay=0.01)
```
Comment (yj0604park) [inline main.py:100]: weight decay가 혹시 너무 커서 학습이 어려운건 아닌지?!

File: main.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    print(f'top 1 acc: {round(correct.item() / len(data), 4)}')
+
+
+def train(model: ClsModel, train_data, optimizer, loss_fn, config):
+    if config['valid']['whether']:
+        print('validation will be proceeded during training')
+        assert config['valid']['steps'] > 0
+
+    if config['valid']['train_split']:
+        print('split data ...')
```
Comment (yj0604park) [inline main.py:52]: 뭐 당장은 복잡하지 않으니 괜찮지만 나중에는 preprocessing 함수를 분리하는게 유지보수에 좋지 않을까 생각해봅니당

File: main.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    # TODO: Modify hard code for optimizer and loss function
+    optimizer = optim.SGD(model.parameters(), lr=config['train']['learning_rate'], momentum=0.9, weight_decay=0.01)
+    loss_fn = nn.CrossEntropyLoss()
+
+    train(model, train_data, optimizer, loss_fn, config)
+
+    if config['eval']['whether']:
+        test_data = datasets.CIFAR100(root=data_dir, train=False, download=True, transform=transforms.Compose(
+            [transforms.ToTensor(), transforms.Normalize(norm['mean'], norm['stdev'])]
```
Comment (yj0604park) [inline main.py:107]: 근데 여기서 Normalize 하는 mean stdev는 train 데이터의 평균값/표준편차값으로 해야할것으로 보이는데, (현재는 그냥 config에 적어 뒀지만) 일반적인 데이터의 경우에는 어떻게 해야할지?! 역시 미리 계산해두고 쓰면 끝이려나

File: conf.d/config.yaml.example (hunk 1 above)
Comment (kimground) [reply conf.d/config.yaml.example:1]: 모델 configuration 관련해서 json보다는 더 쓰이는 거 같긴한데 가능하면 나중에 argparser로 대체해보려고 합니다.

File: conf.d/config.yaml.example (hunk 2 above)
Comment (kimground) [reply conf.d/config.yaml.example:18]: 맞습니다. 더 적절한 변수명이 있다면 제안해주세요!

File: main.py (hunk 3 above)
Comment (kimground) [reply main.py:100]: 헷갈렸는데 원하던 것이 아니어서 성능이 더 잘 나오는 쪽으로 수정해서 반영할 예정입니다.

File: main.py (hunk 4 above)
Comment (kimground) [reply main.py:52]: 넹. data processing 뿐 아니라 모듈화가 필요한 부분은 차후에 수정토록 하겠습니다.

File: main.py (hunk 5 above)
Comment (kimground) [reply main.py:107]: 현재 Normalization에 쓰이고 있는 값들도 CIFAR100의 통계값이 아니라 imagenet의 통계값입니다.
더 자세한 내용은 @taeho-kil 님이 알려주실 수 있을 것 같은데, image normalization에 크게 두 가지 방법이 쓰인답니다.
1. ImageNet의 통계값을 이용하여 normalize
2. 모든 값들이 -1 ~ +1 사이에 있도록 normalize

File: conf.d/config.yaml.example (hunk 2 above)
Comment (yj0604park) [reply conf.d/config.yaml.example:19]: "eval 할지 말지"에서 할지 말지 보다는 eval 자체의 의미를 살리는 변수명이 낫지 않나요?

do_eval이랄지?

Comment (kimground) [reply conf.d/config.yaml.example]: 반영했습니다~

**Review by taeho-kil — APPROVED** [review]
리뷰 승인합니당~

File: main.py (hunk 5 above)
Comment (taeho-kil) [reply main.py]: @kimground 
관례?인지 보통 1번 많이 씁니다. 각 finetune dataset에 대한 statics 구해서 그거로 normalization 잘 안해주더라고요.
video에 대한 action recognition 할 때 2번도 많이 보긴했습니다.
머 benchmark dataset, task 마다 성능 차이가 있긴할텐데 다들 크게 유의하진 않더라고요.

File: main.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    config['device'] = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+
+    norm = {
+        'mean': tuple(map(float, config['data']['norm']['mean'].split())),
+        'stdev': tuple(map(float, config['data']['norm']['stdev'].split()))
+    }
+
+    print('Load data ...')  # TODO: Modify hard code for dataset
+    train_data = datasets.CIFAR100(root=data_dir, train=True, download=True, transform=transforms.Compose(
+        [transforms.ToTensor(), transforms.Normalize(norm['mean'], norm['stdev'])]
```
[the remark this answers is not in the document]
Comment (kimground) [reply main.py]: CIFAR-100 데이터의 image size로 fine-tuning을 진행하니 image token이 4개 밖에 나오지 않아서 학습이 잘되지 않아서 resize를 추가해주었습니다.

File: evaluation.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import torch
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+from typing import Optional
+
+from model import ClsModel
+
+
+def evaluation(model: ClsModel, data, config, valid: Optional[bool] = False):
```
Comment (yj0604park) [inline evaluation.py:9]: evaluation과 validation을 통틀수 있는 단어가 없을까요

File: train.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+from torch import nn, optim
+from torch.utils.data import DataLoader, random_split
+from tqdm import tqdm
+
+from evaluation import evaluation
+from model import ClsModel
+
+
+def train(model: ClsModel, train_data, config):
+    if config['valid']['do_valid']:
```
Comment (yj0604park) [inline train.py:11]: 혹시 python formatter 뭐 쓰시나요?! 없으면 하나 세팅하는게 좋을거 같습니당

File: evaluation.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    model.eval()
+
+    correct = torch.tensor(0, dtype=torch.int, device=config['device'])
+    with torch.no_grad():
+        for images, labels in tqdm(DataLoader(dataset=data, batch_size=config['valid' if valid else 'eval']['batch']),
```
Comment (yj0604park) [inline evaluation.py:14]: valid 라는 boolean 변수값을 false로 줬을때 eval을 한다는건 좀 이상하긴 합니당 (valide = false -> invalid?)
차라리 string으로 입력을 validation_type 받아서 그냥 그걸로 config[validation_type]['batch'] 로 하는게?
물론 validation_type은 'valid' / 'eval' 만 가능하다거나, 대소문자를 다 normalize해서 parameter 값을 한번 체크를 해야하긴 하겠지만요

File: evaluation.py (hunk 7 above)
Comment (kimground) [reply evaluation.py:9]: 일반적으로 어떻게 쓰이는지 찾아보겠습니다. validation과 evaluation이 시점과 데이터만 다르지 결국 코드가 완전히 같아서 합치려는 의도였는데, 몇 개 되지 않지만 그동안 모델 코드에서는 validation 부분이 method로 wrapping되지 않고 train() 안에 그냥 남아있었던 것 같아요.
도와주세요 @taeho-kil 님. 대체로 어떻게 처리하나요?

File: evaluation.py (hunk 9 above)
Comment (kimground) [reply evaluation.py:14]: 음 혹시 그럼 valid 대신에 eval로 값을 받으면 어떨까요? 물론 위의 리뷰가 먼저 해결된 다음에야 방법이 정해질 것 같습니다만

File: train.py (hunk 8 above)
Comment (kimground) [reply train.py:11]: 현재 아예 사용하지 않고 있는데, 해당 코드에서 어떤 점이 formatter 관련 문제인지 알려주실 수 있을까요?
formatter는 찾아보겠습니다.

File: evaluation.py (hunk 9 above)
Comment (yj0604park) [reply evaluation.py:14]: eval로 받는게 덜 헷갈리는 방법일거 같긴 합니다 ㅋㅋ 요새 클린코드 책 공부하는 중인데 거기서 boolean parameter를 까는 부분이 있길래 지적해보았습니다 ㅋㅋ 뭐 익숙해지면 상관없는 부분일거 같긴 해요

Comment (yj0604park) [reply evaluation.py:14]: 여튼 딥러닝 도메인에서 관행이 있으면 그걸 따르는 게 좋을듯 한데 관행이 뭔지를 모르겠네요 ㅋㅋ

File: train.py (hunk 8 above)
Comment (yj0604park) [reply train.py:11]: 아 black 포매터는 "를 쓰기에 black은 안쓰고 계신듯 하여 그냥 한번 체크해봤습니다.

pre-commit에서 체크하도록 해봐도 좋아여

Comment (jongwon-jay-lee) [reply train.py:11]: 매우 딜레이된 참여이지만 저는 flake8, black, isort  썼었어요

File: evaluation.py (hunk 7 above)
Comment (jongwon-jay-lee) [reply evaluation.py:9]: 저는 개인적으로 HF에서 하는 스타일을 좋아합니다.

[def evaluate():](https://github.com/huggingface/transformers/blob/e7cfc1a313cc928e962bb8699868f5dcf46f11eb/examples/question-answering/run_squad.py#L272)

[Training 시 evaluate 호출](https://github.com/huggingface/transformers/blob/e7cfc1a313cc928e962bb8699868f5dcf46f11eb/examples/question-answering/run_squad.py#L235)

[eval 시 evaluate 호출](https://github.com/huggingface/transformers/blob/e7cfc1a313cc928e962bb8699868f5dcf46f11eb/examples/question-answering/run_squad.py#L819)

File: train.py (hunk 8 above)
Comment (kimground) [reply train.py]: 반영했습니다! 이 부분은 pre-commit 추가 PR로 따로 올리겠습니다.

File: evaluation.py (hunk 9 above)
Comment (kimground) [reply evaluation.py]: 위에 리뷰에 달린 종원님 리뷰대로 HF 스타일 따라서 개선해볼게요
